File: etc/[BJ 2150] Strongly Connected Component /Kosaraju.py
import sys
sys.setrecursionlimit(10**9)
scc = []
graph = {}
reverse_graph = {}
stack = []
new_scc = []
def init(v):
    graph[v] = []
    reverse_graph[v] = []
def DFS(v):
    for node in graph[v]:
        if(dfs_visited[node] == False):
            dfs_visited[node] = True
            DFS(node)
    stack.append(v)
    return

def reverse_DFS(v):
    new_scc.append(v)
    for node in reverse_graph[v]:
        if(dfs_visited[node] == False):
            dfs_visited[node] = True
            reverse_DFS(node)

    return
if __name__ == "__main__":
    V, E = map(int, input().split())
    for i in range(1, V+1):
        init(i)
    for i in range(E):
        A, B = map(int, input().split())
        graph[A].append(B)
        reverse_graph[B].append(A)
    dfs_visited = [False]*(V+1)
    for i in range(1, V+1):
        if(dfs_visited[i] == False):
            dfs_visited[i] = True
            DFS(i)
    dfs_visited = [False]*(V+1)
    while(stack):
        vertex = stack.pop()
        if(dfs_visited[vertex] == False):
            new_scc = []
            dfs_visited[vertex] = True
            reverse_DFS(vertex)
            new_scc.sort()
            scc.append(new_scc)
    scc.sort()
    print(len(scc))
    for i in scc:
        for k in i:
            print(k, end = " ")
        print(-1)

File: etc/[BJ 2150] Strongly Connected Component /Kosaraju.cpp
//
//  main.cpp
//  cpp
//
//  Created by 나지혜 on 2021/03/02.
//
#include <string.h>
#include <cstdio>
#include <vector>
#include <stack>
#include <algorithm>
using namespace std;
const int MAX = 100001;
int V, E;
vector<int> adj[MAX];
vector<int> reverse_adj[MAX];
bool visited[MAX];
stack<int> S;
vector<vector<int>> SCC;
vector<int> new_scc;
int Scnum;
void DFS(int v)
{
    for(int next:adj[v])
    {
        if(!visited[next])
        {
            visited[next] = true;
            DFS(next);
        }
    }
    S.push(v);
    return;
}
void reverse_DFS(int v)
{
    new_scc.push_back(v);
    for(int next:reverse_adj[v])
    {
        if(!visited[next])
        {
            visited[next] = true;
            reverse_DFS(next);
        }
    }
    return;
}
int main(){
    scanf("%d %d", &V, &E);
    for (int i=0; i<E; i++){
        int A, B;
        scanf("%d %d", &A, &B);
        adj[A].push_back(B);
        reverse_adj[B].push_back(A);
    }
    for (int i=1; i<=V; i++)
    {
        if(!visited[i])
        {
            visited[i] = true;
            DFS(i);
        }
    }
    memset(visited, false, sizeof(visited));
    while(!S.empty())
    {
        int top = S.top();
        S.pop();
        if(!visited[top])
        {
            new_scc.clear();//모든 원소 삭제
            visited[top] = true;
            reverse_DFS(top);
            sort(new_scc.begin(), new_scc.end());
            SCC.push_back(new_scc);
            Scnum+=1;
        }
        
    }
    sort(SCC.begin(), SCC.end());
    
    printf("%d\n", Scnum);
    for(auto& curr: SCC)
    {
        for(int nowcurr:curr)
            printf("%d ", nowcurr);
        puts("-1");
    }
 
}
